components/ws-manager/pkg/manager/annotations.go

```diff
@@ -145,126 +148,59 @@ func (a *annotation) Apply(dst map[string]string) (needsUpdate bool) {
 }
 
-// patchPodLifecycleIndependentState updates the pod lifecycle independent state of a workspace by setting the
-// non-zero values of the patch. Calling this function triggers a status update. This function is
-// neither atomic, nor synchronized.
-func (m *Manager) patchPodLifecycleIndependentState(ctx context.Context, workspaceID string, patch func(*podLifecycleIndependentState) (needsUpdate bool), annotations ...*annotation) (err error) {
-	//nolint:ineffassign
-	span, ctx := tracing.FromContext(ctx, "patchPodLifecycleIndependentState")
-	defer tracing.FinishSpan(span, &err)
-
-	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
-		ctx, cancel := context.WithTimeout(ctx, kubernetesOperationTimeout)
-		defer cancel()
-
-		var plisCfg corev1.ConfigMap
-		err := m.Clientset.Get(ctx, types.NamespacedName{Namespace: m.Config.Namespace, Name: getPodLifecycleIndependentCfgMapName(workspaceID)}, &plisCfg)
-		if isKubernetesObjNotFoundError(err) {
-			return xerrors.Errorf("workspace %s has no pod lifecycle independent state", workspaceID)
-		}
+func (m *Manager) modifyFinalizer(ctx context.Context, workspaceID string, finalizer string, add bool) error {
+	// Retry on failure. Sometimes this doesn't work because of concurrent modification. The Kuberentes way is to just try again after waiting a bit.
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+		span, ctx := tracing.FromContext(ctx, "modifyFinalizer")
+		tracing.ApplyOWI(span, log.OWI("", "", workspaceID))
+		defer tracing.FinishSpan(span, &err)
+		span.LogKV("finalizer", finalizer, "add", add)
+
+		pod, err := m.findWorkspacePod(ctx, workspaceID)
 		if err != nil {
-			return xerrors.Errorf("cannot retrieve pod lifecycle independent state: %w", err)
+			return xerrors.Errorf("cannot find workspace %s: %w", workspaceID, err)
+		}
+		if pod == nil {
+			return xerrors.Errorf("workspace %s does not exist", workspaceID)
 		}
-		tracing.LogEvent(span, "k8s get done")
 
-		needsUpdate := false
-		if patch != nil {
-			plis, err := unmarshalPodLifecycleIndependentState(&plisCfg)
-			if err != nil {
-				return xerrors.Errorf("patch pod lifecycle independent state: %w", err)
+		var update bool
+		if add {
+			var exists bool
+			for _, x := range pod.Finalizers {
+				if x == gitpodFinalizerName {
+					exists = true
+					break
+				}
 			}
-			if plis == nil {
-				plis = &podLifecycleIndependentState{}
+			if !exists {
+				pod.Finalizers = append(pod.Finalizers, finalizer)
+				update = true
 			}
-			tracing.LogEvent(span, "unmarshalling done")
-			tracing.LogKV(span, "prePatchPLIS", plisCfg.Annotations[plisDataAnnotation])
-
-			needsUpdate = patch(plis)
-			tracing.LogEvent(span, "patch done")
-
-			err = marshalPodLifecycleIndependentState(&plisCfg, plis)
-			if err != nil {
-				return xerrors.Errorf("patch lifecycle independent state: %w", err)
+		} else {
+			n := 0
+			for _, x := range pod.Finalizers {
```

Comment:
This loops feels a tad to nifty.
 - Why the explicit (re-)assignment of finalizers (in the loop and later: `pod.Finalizers = pod.Finalizers[:n]`)? If we skipped that the code would become easier
Is there a Kubernetes guarantee that there is ever only one entry with a given name?

Reply:
I simply applied the filter slice trick: https://github.com/golang/go/wiki/SliceTricks#filter-in-place :)

Reply:
Yes, I think I got the effect. Would have expected sth like `update := len(pod.Finalizers) != len(filteredFinalizers)` for better readability. But your mileage may vars :)